docs/src/nondiff_points.md

```diff
@@ -0,0 +1,141 @@
+using Plots
+gr(framestyle=:origin, legend=false)
+```
+### Derivative is defined in usual sense
+```@example nondiff
+plot(x->x^3)
+```
+This is the standard case, one can returned the derivative that is defined according to school room calculus.
+Here we would reasonably say that at `x=0` the derivative is `3*0^2=0`. 
+
+
+An interesting thing about `x->x^3` is that at `x=0` the derivative is defined, but neither the sub-derivative nor super-derivative is defined.
+
+### Local Minima / Maxima
+
+```@example nondiff
+plot(abs)
+```
+
+`abs` is the classic example of a function where the derivative is not defines as the limit from above is not equal to the limit from below
+
+$$\mathrm{abs}'(0) = \lim_{h->0^-} \dfrac{\mathrm{abs}(0)-\mathrm{abs}(0-h)}{0-h} = -1$$
+$$\mathrm{abs}'(0) = \lim_{h->0^+} \dfrac{abs(0)-\mathrm{abs}(0-h)}{0-h} = 1$$
+
+Now, as discussed in the introduction the AD system would on it's own choose either 1 or -1, depending on implementation.
+
+We however have a potentially much nicer answer available to use: 0.
+
+This has a number of advantages.
+- It follows the rule that derivatives are zero at local minima (and maxima).
+- If you leave a gradient decent optimizer running it will eventually actually converge absolutely to the point -- where as with it being 1 or -1 it would never outright converge it would always flee.
+
+Further:
+- It is a perfectly nice member of the [subderivative](https://en.wikipedia.org/wiki/Subderivative).
+- It is the mean of the derivative on each side; which means that it will agree with central finite differencing at the point.
+### Piecewise slope change
+```@example nondiff
+plot(x-> x < 0 ? x : 5x)
+```
+
+Here was have 3 main options, all are good.
+
+We could say there derivative at 0 is:
+ - 1: which agrees with backwards finite differencing
+ - 5: which agrees with forwards finite differencing
+ - 3: which is the mean of `[1, 5]`, and agrees with central finite differencing
+
+All of these options are perfectly nice members of the [subderivative](https://en.wikipedia.org/wiki/Subderivative).
+`3` is the arguably the nicest, but it is also the most expensive to compute.
+In general all are acceptable.
+
+
+### Derivative zero almost everywhere
+
+```@example nondiff
+plot(ceil)
+```
+
+Here it is most useful to say the derivative is zero everywhere.
+The limits are zero from both sides.
+
+The other option for `x->ceil(x)` would be relax the problem into `x->x`, and thus  say it is 1 everywhere
+But that it too weird, if the use wanted a relaxation of the problem then they would provide one.
+We can not be imposing that relaxation on to `ceil` for everyone is not reasonable.
+
+### Not defined on one-side
+```@example nondiff
+plot(x->exp(2log(x)))
+plot!(; xlims=(-10,10), ylims=(-10,10)) #hide
+```
+
+We do not have to worry about what to return for the side where it is not defined.
+As we will never be asked for the derivative at e.g. `x=-2.5` since the primal function errors.
```

Comment:
Just a comment, not sure if it's important but the primal won't error if we make the argument complex. And in that case there's the interesting issue of the branch cut.